### ISSUE TITLE
Realtime + RLS not working @ hosted supabase cloud

### ISSUE THREAD

**Issue by skizer** [supabase/realtime#265]
# Bug report

## Describe the bug

As with 09.07.2022 realtime in combination with activated RLS seem's to be broken with a freshly created supabase cloud instance.

Socket get's created, connects, schema realtime->subscription gets populated with the correct token/role.
Work's fine if RLS is turned off for the corresponding table, stop's working the moment RLS is turned on.

All permissions have been granted.

~~P.S This hasn't been tested with a self-hosted instance but only the cloud instance hosted directly by Supabase.~~
Seem's to work fine one a self-hosted instance

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Setup a new hosted supabase cloud instance
2. Setup a new table, enable RLS and replication for that specific table
4. Grant ALL permission for testing purpose to PUBLIC
5. Test it, won't work.
6. Deactivate RLS -> works

## System information

- OS: Debian GNU/Linux 11 (bullseye)
- Browser: Chrome
- Version of supabase-js: 1.35.4
- Version of Node.js: v16.15.1

## Additional context

- This has been discovered by retesting and reproducing the exact same setup from @GaryAustin1. [#7630](https://github.com/supabase/supabase/discussions/7630)
- This has also been tested on a 6 months old cloud instance, working as expected, no problems at all.

**Comment by skizer** [comment]
A little update @w3b6x9 , out of curiosity I've just setup a fresh self-hosted supabase instance with docker, pulled right from supabase/docker - Guess what, it works fine there.

**Comment by mrolafsson** [comment]
I'm seeing the same. RLS works for regular `select` but not for realtime. Disabling RLS push works as expected. My policy uses a function in case that helps.

**Comment by w3b6x9** [comment]
@skizer it might be related to a few migrations that are on the self-hosted supabase instance docker that hasn't made its way yet to the hosted version. i was going to deploy some fix changes to the entire supabase fleet tomorrow afternoon but let me do it for your project now to see if it resolves your issue.

**Comment by w3b6x9** [comment]
@mrolafsson would you like me to do it for your project as well?

**Comment by w3b6x9** [comment]
@skizer @mrolafsson it might have to do with a specific type of a column and it could be a bug on our end. i'll send this to the rest of the team and we'll investigate fully. thanks for your patience!

**Comment by w3b6x9** [comment]
@skizer @mrolafsson can both of you please try again in your respective projects?

@skizer i manually tested your project's realtime and it was working for me. if it's still not working for you make sure you are passing in the right apikey/token on subscription and that your rls policy/policies work as intended.

**Comment by w3b6x9** [comment]
@skizer @mrolafsson this is indeed a bug on our end introduced in one of the latest migrations. we just created a fix and will push out a fix ASAP.

**Comment by w3b6x9** [comment]
If you want to get unblocked ASAP you can manually run this query: https://github.com/supabase/walrus/blob/4eb365250347787a935c21ce9790d2307a17b379/sql/walrus--0.1.sql#L204-L236

**Comment by conradlin** [comment]
Hi guys, appreciate all your hard work. I made a comment in the discord, and was recommended to share my details here - my realtime subscribes in my projects are not triggering anymore, even when there is definitely a change in the database. No configuration has changed for the last few days, and only noticing this problem today.

I have also rebooted my project, and still no luck. My project URL is https://jdchgovbmzfgodhranzg.supabase.co and my code is all open source here: https://github.com/thex3family/make-work-fun/

Do let me know what I can do on my end to be helpful to resolve this problem. Thank you!

**Comment by w3b6x9** [comment]
@conradlin your issue is unrelated but i have resolved it so realtime should be working for your project now.

**Comment by skizer** [comment]
> @skizer @mrolafsson can both of you please try again in your respective projects?
> 
> @skizer i manually tested your project's realtime and it was working for me. if it's still not working for you make sure you are passing in the right apikey/token on subscription and that your rls policy/policies work as intended.

Hey,
thanks for your feedback!

I've tested it and it didn't work right away so I had to run your mentioned query but once done, things started to work.
Dunno if it helps but I thought you should know that since you mentioned you ran the query already on my project.

**Comment by mrolafsson** [comment]
> If you want to get unblocked ASAP you can manually run this query: https://github.com/supabase/walrus/blob/4eb365250347787a935c21ce9790d2307a17b379/sql/walrus--0.1.sql#L204-L236

Works perfectly!!! Thank you so much for all your help!

**Comment by mitchemmc** [comment]
> If you want to get unblocked ASAP you can manually run this query: https://github.com/supabase/walrus/blob/4eb365250347787a935c21ce9790d2307a17b379/sql/walrus--0.1.sql#L204-L236

We ran this and it fixed the issue with all events except the delete event. We now have an issue where delete events ignore the row level filtering `{table}:{col}=eq.{val}` and now every user is being updated for all delete events no matter the filter used.

This also seems to be a big security vulnrability for us as all users see the entire row for every delete event (because we need full data replication). I can provide more details if needed.

**Comment by w3b6x9** [comment]
@mitchemmc I think it's our fault for not properly documenting this but delete events are a bit different:

> Important Notes:
> 
> - Row level security is not applied to delete statements
> - The key/value pairs displayed in the old_record field include the table's identity columns for the record being updated/deleted. To display all values in old_record set the replica identity for the table to full
> - When a delete occurs, the contents of old_record will be broadcast to all subscribers to that table so ensure that each table's replica identity only contains information that is safe to expose publicly

- https://github.com/supabase/walrus

**Comment by mitchemmc** [comment]
> @mitchemmc I think it's our fault for not properly documenting this but delete events are a bit different:
> 
> > Important Notes:
> > 
> > * Row level security is not applied to delete statements
> > * The key/value pairs displayed in the old_record field include the table's identity columns for the record being updated/deleted. To display all values in old_record set the replica identity for the table to full
> > * When a delete occurs, the contents of old_record will be broadcast to all subscribers to that table so ensure that each table's replica identity only contains information that is safe to expose publicly
> 
> * https://github.com/supabase/walrus

Thanks for the info.

I have a database that I set up a few months ago and when I run this command:

```
supabase
  .from(`table:id=eq.1`)
  .on("DELETE", (payload) => {
    console.log(payload);
  })
  .subscribe();
```

I will receive delete events only on the row with the ID of 1.
However on my database that I created last week the same query will trigger for every row in the database.
So it is also ignoring my filters on delete events which seems to be a recent change.

I've also stopped listening to delete events entirely (only listening for updates) and if I check my network tab in chrome dev tools I still see the delete events coming in to the websocket.

This seems to be a bug to me as the new database I created that has this behaviour has significantly worse performance compared to the database I created a few months ago and the only difference I can see is that there are way more realtime events because of this behaviour change.

I've recently created a support ticket for this so if GitHub issues isn't the best place to discuss this feel free to pick it up there.

**Comment by w3b6x9** [comment]
@mitchemmc thanks so much for the additional info. We'll take a deeper look and try to uncover what's causing this discrepancy.

**Comment by w3b6x9** [comment]
Update: the original issue here has been resolved and the fix has gone out to all Supabase projects. If you continue to face the originally reported issue then please drop a comment here. Otherwise, please create other GitHub issues so that we may address them individually.

**Comment by steffenstolze** [comment]
We've encountered the exact same problem, trying to use realtime with RLS and a pretty basic use case on hosted Supabase.

Realtime updates we're not triggered until we ran the following statement:

```sql
create or replace function realtime.build_prepared_statement_sql(
    prepared_statement_name text,
    entity regclass,
    columns realtime.wal_column[]
)
    returns text
    language sql
as $$
/*
Builds a sql string that, if executed, creates a prepared statement to
tests retrive a row from *entity* by its primary key columns.
Example
    select realtime.build_prepared_statement_sql('public.notes', '{"id"}'::text[], '{"bigint"}'::text[])
*/
    select
'prepare ' || prepared_statement_name || ' as
    select
        exists(
            select
                1
            from
                ' || entity || '
            where
                ' || string_agg(quote_ident(pkc.name) || '=' || quote_nullable(pkc.value #>> '{}') , ' and ') || '
        )'
    from
        unnest(columns) pkc
    where
        pkc.is_pkey
    group by
        entity
$$;
```

No it works. So.... what's the deal with this? Will it be fixed? Will we have to execute that statement in every new project?

Thank you!

**Comment by w3b6x9** [comment]
@steffenstolze so it's been fixed in Realtime (see: https://github.com/supabase/realtime/pull/266) but we haven't deployed it yet for all projects. we'll do this as soon as we can!

**Comment by mathewcst** [comment]
any updates on this? I'm trying to figure out why my INSERTs where not reflected for a week now

**Comment by mathewcst** [comment]
I'm trying to figure Realtime out for a week now, and only deletes were working. 

I'm using Prisma + Sveltekit with Supabase. In a recent test, I've discovered that the problem was using tables created by prisma. If I create at able manually (on supabase dashboard), it works.

Any ideas why?

**Comment by GaryAustin1** [comment]
> I'm trying to figure Realtime out for a week now, and only deletes were working.
> 
> I'm using Prisma + Sveltekit with Supabase. In a recent test, I've discovered that the problem was using tables created by prisma. If I create a table manually (on supabase dashboard), it works.
> 
> Any ideas why?

Probably this:  https://supabase.com/docs/guides/integrations/prisma#troubleshooting
Prisma messes up grants.

**Comment by mathewcst** [comment]
Unfortunately not. Tried this already and had no success on inserts, only deletes.

**Comment by GaryAustin1** [comment]
Deletes don't use RLS.  Inserts and Updates do.  Do you have RLS on and do you meet the select policy?  Does it work with RLS off?

**Comment by mathewcst** [comment]
Tried both with RLS on and off, but no luck.

Since I use Prisma on the "backend" (Sveltekit Endpoints), it actually inserts on the database but doesn't reflect on the subscription.

**Comment by w3b6x9** [comment]
@mathewcst it's probably related to Prisma creating quoted custom types on your behalf. There's a fix but waiting to be deployed to all projects. Can you please create a ticket at https://app.supabase.com/support/new and we'll look into it?

**Comment by mathewcst** [comment]
@w3b6x9 will do. Ty.

**Comment by pie6k** [comment]
I just want to express that this issue is yet another one that makes our team want to drop using supabase altogether.

We're willing to pay for supabase cloud easily, as it provides us a lot of value.

But it feels very 'early beta' in so many places it is actually a bit cringe.

- it is necessary to run some magic SQL queries, which are not version-controlled
- without that real-time is simply not working without any warnings or errors
- there are no announcements about it in dashboard, meaning it is on me or some other dev finding this github issue and magical SQL snippets

We'll still try as we want to move fast and supabase seemed like a good fit. But it starts to feel more and more like using an early beta, not a production-ready product.

**Comment by mathewcst** [comment]
I have to say I felt the same way as you @pie6k when I was using Prisma. After learning that Prisma changed a bit of my schema/permissions, I just dropped it, and everything just kinda worked (thanks to the Supabase support team).

Maybe contacting the support and explaining what's the problem for you, can help fix it, they're super quick on the responses.

**Comment by w3b6x9** [comment]
Completely understand, @pie6k!

> Not sure if this is the best place for such a comment.

This is definitely the right place - we encourage feedback on any channel. We definitely could have picked up on some of these problems with a more robust test suite - it's always hard to balance building fast and building right. 

The immediate action we can take here is to increase our test coverage. We already discussed this in our Launch Week retrospective yesterday, and agreed as a company that it would be a good thing. 

> We'll still try as we want to move fast and supabase seemed like a good fit

Thank you - we'll make sure you don't regret it.

**Comment by ass77** [comment]
> Unfortunately not. Tried this already and had no success on inserts, only deletes.

 Im using prisma ORM on my express server and I've encountered the same issues, `INSERT` and `UPDATE` doesnt work as expected, only `DELETE`.

p/s: I didn't use any `RLS` on my tables.

I've been following these steps to troubleshoot it, but the problem still persists:


1. everytime I migrate my db, i ran this `SQL` command: https://supabase.com/docs/guides/integrations/prisma#resources
2. injected this `SQL` command to unblock realtime asap: https://github.com/supabase/walrus/blob/4eb365250347787a935c21ce9790d2307a17b379/sql/walrus--0.1.sql#L204-L236
3. also this, changing the types not starting with capital letter in prisma schema: https://github.com/supabase/supabase/issues/5685#issuecomment-1054703357

**Comment by w3b6x9** [comment]
@ass77 are you self-hosting Realtime or using a Supabase project? If you're at liberty, can you share your exact setup steps? Otherwise, can you create a support ticket at https://app.supabase.com/support/new.

**Comment by ass77** [comment]
@w3b6x9 

Im using supabase project (cloud).  I've found the issue actually, seems like when I use `enum` on prisma schema, it will not work. 

Then, I need to destroy the db and push my schema again (without enum) for it to work properly.

This is the exact setup step with the issue:

1. create prisma schema with `enum` fields - can
2. push schema using `npx prisma db push`
3. migrate schema using `npx prisma migrate dev --name init`
4. create client using `npx prisma generate` for CRUD operation
5. run LISTEN & NOTIFY server using supabase realtime like this - https://github.com/supabase/realtime/blob/master/examples/node-js/src/server.js

**Comment by LivioGama** [comment]
This can also be due to this if you have trouble with realtime, upgrade supabase client to 2.24.0 https://github.com/supabase/supabase-js/commit/b55d386b35cdf9373fabbbc07199f86811bde339

**Comment by LivioGama** [comment]
That was not it, I decided to write a whole medium post for avoiding you waste time in the future: https://liviogamassia.medium.com/using-supabase-rls-with-firebase-auth-custom-auth-provider-357eaad9c70f

**Comment by clajps0n** [comment]
When I was about to give up, but, in the Database menu, I found a section called Publications (previously called Replications). The only thing I did was enable each of my tables and IT WORKED!!!

Ignorance causes headaches:

Infrastructure Setup: Enabling real-time sets up the necessary infrastructure and WebSocket connections for real-time updates.

Selective Change Tracking: Publications ensure that only the specified changes are tracked and sent over the WebSocket connections. Without Publications, PostgreSQL wouldn't know which changes to replicate, even if the infrastructure for real-time updates is in place.

I hope this information is useful for you too.
